### PR TITLE
Extend GitRepoTests to allow include mode

### PR DESCRIPTION
**Pull request by kewillford**
The GitRepoTests can be initialized with a ValidateWorkingTree
mode that is currently a boolean. If enabled, we will add extra
checks to make sure the working tree of our virtual enlistment
matches the enlistment from a vanilla Git repo.

To extend this to the "include" mode, we need to first modify
this boolean into an enum and add it as an extra option.

Second, we need to be very careful about what we are comparing.
We will no longer virtualize outside of the included folders.
Be careful to include files in the root directory.

The functional tests can then be run with "--partial-mode" to
exercise only this option, which will isolate the test failures
to this setting AND keep the test suite from getting too slow.

Currently, the partial mode space is probably wider than
necessary and can be reduced in later iterations. This
does provide _some_ coverage.

Signed-off-by: Derrick Stolee <dstolee@microsoft.com>

### CODE REVIEW

File: GVFS/GVFS.FunctionalTests/Should/FileSystemShouldExtensions.cs [hunk 1]
```diff
@@ -251,12 +252,56 @@ public DirectoryInfo WithAttribute(FileAttributes attribute)
...
+            private static bool IsMatchedPath(FileSystemInfo info, string repoRoot, HashSet<string> prefixes)
+            {
+                if (prefixes == null)
+                {
+                    return true;
+                }
+
+                string localPath = info.FullName.Substring(repoRoot.Length + 1);
+
+                if (localPath.Equals(".git"))
```
Comment (wilbaker) [inline GVFS/GVFS.FunctionalTests/Should/FileSystemShouldExtensions.cs]: Should this also be `OrdinalIgnoreCase`?

File: GVFS/GVFS.FunctionalTests/Should/FileSystemShouldExtensions.cs [hunk 2]
```diff
@@ -297,24 +343,27 @@ public DirectoryInfo WithAttribute(FileAttributes attribute)
...
                             {
                                 expectedMoved = expectedEnumerator.MoveNext();
+
                                 continue;
                             }
                         }
 
                         Assert.Fail($"File names don't match: expected: {expectedEnumerator.Current.FullName} actual: {actualEnumerator.Current.FullName}");
+                        expectedMoved = expectedEnumerator.MoveNext();
+                        continue;
```
Comment (wilbaker) [inline GVFS/GVFS.FunctionalTests/Should/FileSystemShouldExtensions.cs]: Why was this change needed?  With the call to `Assert.Fail` on line 351 will we ever hit this code?

File: GVFS/GVFS.FunctionalTests/Should/FileSystemShouldExtensions.cs [hunk 3]
```diff
@@ -297,24 +343,27 @@ public DirectoryInfo WithAttribute(FileAttributes attribute)
...
                     }
                     else
                     {
-                        (actualEnumerator.Current.Attributes & FileAttributes.Directory).ShouldNotEqual(FileAttributes.Directory, $"expected path: {expectedEnumerator.Current.FullName} actual path: {actualEnumerator.Current.FullName}");
+                        (actualEnumerator.Current.Attributes & FileAttributes.Directory).ShouldNotEqual(FileAttributes.Directory, $"expected file path: {expectedEnumerator.Current.FullName} actual directory path: {actualEnumerator.Current.FullName}");
 
                         FileInfo expectedFileInfo = (expectedEnumerator.Current as FileInfo).ShouldNotBeNull();
                         FileInfo actualFileInfo = (actualEnumerator.Current as FileInfo).ShouldNotBeNull();
-                        actualFileInfo.Length.ShouldEqual(expectedFileInfo.Length);
+                        actualFileInfo.Length.ShouldEqual(expectedFileInfo.Length, $"File lengths do not agree for {expectedEnumerator.Current.FullName} vs {actualEnumerator.Current.FullName}");
```
Comment (wilbaker) [inline GVFS/GVFS.FunctionalTests/Should/FileSystemShouldExtensions.cs]: Include the lengths in the message as well?  (I can't remember if the default message makes it clear which values are expected and actual.)

File: GVFS/GVFS.FunctionalTests/Should/FileSystemShouldExtensions.cs (hunk 2 above)
Comment (kewillford) [reply GVFS/GVFS.FunctionalTests/Should/FileSystemShouldExtensions.cs]: Not sure why @derrickstolee added this.  Seems unnecessary but maybe debugging you can continue through the `Fail`?

Comment (derrickstolee) [reply GVFS/GVFS.FunctionalTests/Should/FileSystemShouldExtensions.cs]: I think I had changed the assert to be something that concatenated the error (so I could see the full list of differences) but somehow forgot to revert this change.

Comment (derrickstolee) [reply GVFS/GVFS.FunctionalTests/Should/FileSystemShouldExtensions.cs]: I think I had changed the assert to be something that concatenated the error (so I could see the full list of differences) but somehow forgot to revert this change.

File: GVFS/GVFS.FunctionalTests/Program.cs [hunk 4]
```diff
@@ -56,10 +63,18 @@ public static void Main(string[] args)
             }
             else
             {
+                Settings.ValidateWorkingTreeMode validateMode = Settings.ValidateWorkingTreeMode.Full;
+
+                if (runner.HasCustomArg("--partial-mode"))
```
Comment (derrickstolee) [inline GVFS/GVFS.FunctionalTests/Program.cs]: To be consistent, should we rename this '--include-mode"?

File: GVFS/GVFS.FunctionalTests/Should/FileSystemShouldExtensions.cs [hunk 5]
```diff
@@ -251,12 +252,54 @@ public DirectoryInfo WithAttribute(FileAttributes attribute)
...
+                {
+                    if (localPath.StartsWith(prefixDir, StringComparison.OrdinalIgnoreCase))
+                    {
+                        return true;
+                    }
+
+                    if (prefixDir.StartsWith(localPath, StringComparison.OrdinalIgnoreCase) &&
+                        Directory.Exists(info.FullName))
+                    {
+                        // For example: localPath = "GVFS" and prefix is "GVFS\\GVFS".
```
Comment (wilbaker) [inline GVFS/GVFS.FunctionalTests/Should/FileSystemShouldExtensions.cs]: > prefix is "GVFS\\GVFS"

Do we have an prefixes in our test that are more than one folder deep?

I might have missed it, but as best I could tell everything in `IncludeModeFolders` is a top level folder.

Comment (kewillford) [reply GVFS/GVFS.FunctionalTests/Should/FileSystemShouldExtensions.cs]: Sorry this was my fault.  I didn't pull it [this commit](https://github.com/derrickstolee/VFSForGit/commit/c10e8ba2bf0ba53c6add2749f83f1928169c79b9)
